### PR TITLE
[Fix] Apply repetition penalty once through SGLang for all shared-sampler models

### LINKED ISSUE #1360: [Bug] repetition_penalty is applied twice (effective penalty is squared)
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang-omni/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

`ModelRunner._sample_next_token_ids` applies `repetition_penalty` manually (`sglang_omni/model_runner/base.py`):

```python
def _sample_next_token_ids(self, logits_output, forward_batch, schedule_batch, requests):
    self._apply_repetition_penalty(logits_output, requests)      # <-- also done by SGLang
    self._apply_codec_suppress_tokens(logits_output, requests)
    self._install_sampling_seeds(forward_batch, requests)
    ...
    next_token_ids = self.tp_worker.model_runner.sample(logits_output, forward_batch)
```

SGLang already applies the same rule through `BatchedRepetitionPenalizer`, reached from `SamplingBatchInfo.apply_logits_bias` inside `tp_worker.model_runner.sample()`. Both run on the same logits, so the configured penalty is applied twice: **`repetition_penalty=2.0` behaves as 4.0.**

Measured on a single logit of an already-generated token:

```
with    the base.py call:  +6.5237 -> +1.6309   (4.00x)
without the base.py call:  +6.5237 -> +3.2619   (2.00x)
```

Both stages read the same `req.sampling_params.repetition_penalty` and act on the same token set, and `_sample_next_token_ids` always routes through `apply_logits_bias`, so there is no configuration in which only one of them runs. Confirmed at runtime: `penalizer_orchestrator.is_required = True`, and the number of entries SGLang modifies grows by one per decode step, matching `len(req.output_ids)`.

Separately, `_apply_repetition_penalty` rebuilds the full `(row, token, penalty)` list every decode step (`O(batch × output_len)`) and materialises it with `torch.tensor(python_list, device="cuda")` three times — a blocking pageable H2D copy on the scheduler thread. In our setup it accounted for ~59% of scheduler-thread CPU time and starved the GPU (SM utilisation averaged 73%, never reaching 90%); removing the call raised end-to-end throughput 8–13% across concurrency 1–64. That is a side effect, not the reason to fix it.

### Reproduction

Snapshot the logits on entry to `_sample_next_token_ids` (before either stage) and compare against the value after SGLang's `apply_logits_bias`. The snapshot must be taken outside `_apply_repetition_penalty` so the script still reports when that call is removed.

```python
import torch
from sglang.srt.sampling.sampling_batch_info import SamplingBatchInfo
from sglang_omni.model_runner.base import ModelRunner

snap, n = {}, [0]

_sample = ModelRunner._sample_next_token_ids
def traced_sample(self, logits_output, forward_batch, schedule_batch, requests):
    snap["raw"] = logits_output.next_token_logits.clone()      # before any shaping
    return _sample(self, logits_output, forward_batch, schedule_batch, requests)
ModelRunner._sample_next_token_ids = traced_sample

_bias = SamplingBatchInfo.apply_logits_bias
def traced_bias(self, logits):
    _bias(self, logits)                                        # after SGLang's pass
    if "raw" in snap and n[0] < 4:
        changed = (snap["raw"] != logits)[0].nonzero().flatten()
        if len(changed):
            t = int(changed[0])
            raw, final = snap["raw"][0, t].item(), logits[0, t].item()
            print(f"token {t}: raw {raw:+.4f} -> final {final:+.4f} = {raw / final:.2f}x")
            n[0] += 1
SamplingBatchInfo.apply_logits_bias = traced_bias
```

Import this before the engine starts, then run any decode with `repetition_penalty=2.0`. Greedy sampling (`temperature=0`) keeps the trace stable, and the ratio is what matters — the specific token ids and logit values vary with the model and sampling settings.

Observed:

```
token 33: raw +6.5237 -> final +1.6309 = 4.00x     # expected 2.00x
token 10: raw +5.8992 -> final +1.4748 = 4.00x
```

After deleting the `self._apply_repetition_penalty(...)` line, the same script prints:

```
token 33: raw +6.5237 -> final +3.2619 = 2.00x
token 10: raw +5.8992 -> final +2.9496 = 2.00x
```

### Environment

- SGLang 0.5.12.post1
- PyTorch 2.11.0+cu130, CUDA 13.0, Python 3.12
- NVIDIA L40S (46 GB), single GPU, TP 1
- overlap schedule disabled, FlashInfer attention backend, async decode enabled

### PR DESCRIPTION
**Pull request by lijrjyan**
## Motivation

On the shared Omni sampling path both Omni's model runner and SGLang's sampler apply `repetition_penalty` to the logits, so the effective penalty is squared and regular decode and speculative verify disagree on the same prefix. 

Apply repetition penalty once through SGLang for all shared-sampler models, so that the parameter keeps its documented strength and every sampling path produces the same logits.
## Modifications

- **`model_runner/base.py`: apply the penalty only in SGLang's sampler and keep its history across retraction.**
  - Remove Omni's repetition-penalty pass and its token-history cache, leaving repetition/frequency/presence penalties to SGLang's penalizer; Omni keeps only codec suppress tokens.
  - On extend batches, re-seed the retained output history into the prepared penalizers before the forward sampling snapshot, so a request re-prefilled after retraction keeps its penalty state for every model on this path.
- **Remove per-model workarounds and set values for single application.**
  - Qwen3-TTS: drop its private restore and no-op override, now covered by the base runner.
  - Fish Audio S2-Pro: hand SGLang `repetition_penalty=1.0`, since the family penalizes in-model.
  - Fun-CosyVoice3: set the default `repetition_penalty` to 1.21, the strength the old 1.1 default was effectively served at, so default output is unchanged.
- **Tests and docs**: add tests to pin sampler ownership, penalty-history restore and the Fish Audio S2-Pro, talker and Fun-CosyVoice3 builder values; remove the obsolete Omni-penalty tests; update the Fun-CosyVoice3 cookbook with the new default.

## Related Issues

Fixes #1360. Supersedes #1817; extends #1750 (Qwen3-TTS only) to every model on the shared sampler.

## Accuracy Test

Measured on H200 with the default configuration (CUDA graphs on): main at `1b801a7c` vs main with this PR.

### Generation quality

* The generated outputs are identical between the main branch and this PR across all tested cases:
  * **Qwen3-TTS 0.6B** with 40 prompts
  * **Fish Audio S2-Pro** with 12 prompts
  * **MOSS-TTS v1.5** with 12 prompts
  * **Whisper large-v3** with 22 clips
  * **Qwen3-Omni** thinker with 6 prompts (greedy decoding); the talker was compared with CUDA graphs off (its graph-mode decode is not run-to-run deterministic even with top_k=1)

* **Fun-CosyVoice3** with 40 prompts (T0.7): 27/40 token streams identical; en WER 0.010 vs 0.010, zh CER 0.0243 vs 0.0243, 0 collapses, mean codes 121.7 (main) vs 120.6 (this PR).

### Penalty semantics

* **Qwen3-Omni thinker** with 6 prompts (greedy), explicit `repetition_penalty`: main at 1.1 and this PR at 1.21 produce identical text on 6/6, confirming main served the square.
* **Fun-CosyVoice3** with 40 prompts: omitting the sampling fields and passing them explicitly (rp 1.21, T0.7, top_p 0.8, top_k 20) give identical output on 40/40.

### Retraction

* **Qwen3-TTS 0.6B** with 40 prompts at concurrency 4, forced retraction every 16 decode steps (`SGLANG_TEST_RETRACT`): 29 requests were retracted and re-prefilled, their penalty state was restored exactly (47/47 re-prefilled rows), and with CUDA graphs off the codec tokens match the no-retraction run on 40/40.
* **Fun-CosyVoice3** with 6 prompts at concurrency 8: re-prefill restores the full penalty history (15/15 tokens; main restores none). The re-prefilled requests themselves cannot complete on main or on this PR because of a pre-existing Fun-CosyVoice3 bug (the RoPE kernel rejects the re-prefill batch), so the end-to-end retraction check is the Qwen3-TTS run above.

Behaviour change: an explicit `repetition_penalty` on the shared path now has its documented strength instead of its square.

## Benchmark & Profiling

Not applicable.

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` or `/tag-and-rerun-ci qwen3-tts` to select a TTS CI
model, and
`/tag-and-rerun-ci fun-asr`, `/tag-and-rerun-ci qwen3-asr` or
`/tag-and-rerun-ci whisper-asr` to select an ASR CI model. One selector from
each family can be combined, for example `/tag-and-rerun-ci moss fun-asr`.
Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/fishaudio_s2_pro/request_builders.py [hunk 1]
```diff
@@ -99,39 +97,24 @@ def build_sglang_tts_request(
...
+    vocab_size = len(tokenizer)
 
     sampling_params = SamplingParams(
         max_new_tokens=state.max_new_tokens,
         temperature=state.temperature,
         top_p=state.top_p,
         top_k=state.top_k,
-        repetition_penalty=state.repetition_penalty,
+        # note (Junnan Li): the in-model sampler already penalizes semantic history.
+        repetition_penalty=1.0,
```
Comment (AkazaAkane) [inline sglang_omni/models/fishaudio_s2_pro/request_builders.py:135]: Fish S2-Pro still depends on output history for its in-model repetition/RAS sampling, but this change sets SGLang `repetition_penalty=1.0`. That makes the base `lookahead_eligible()` treat Fish as history-free and can incorrectly enable async lookahead.

Can we explicitly keep Fish out of lookahead for now, e.g. override `lookahead_eligible()` to return `False`, and add a regression test for this?

Comment (lijrjyan) [reply sglang_omni/models/fishaudio_s2_pro/request_builders.py:135]: Done. Explictly keep fish out of lookahead by overriding `lookahead_eligible()` and add corresponding test. btw, no sure if we had give it a try to split Fish's decode into launch/resolve in order to enable lookahead for fish.

**Review by AkazaAkane — APPROVED** [review]
LGTM code-wise; run the intended GPU CI once, then merge.

File: sglang_omni/models/fishaudio_s2_pro/request_builders.py (hunk 1 above)
Comment (AkazaAkane) [reply sglang_omni/models/fishaudio_s2_pro/request_builders.py:135]: Yeah, I think that could be a follow-up optimization. It needs a bit more than just splitting post_decode, though: Fish currently updates both the semantic/RAS history and last_codebook_values during collect/resolve, and both are inputs to the next decode step.
